### ISSUE TITLE
Nothing works on the screen  on some android devices if I will try to use this component.

### ISSUE THREAD

**Issue by beqramo** [gorhom/react-native-bottom-sheet#566]
# Bug
Hi,
I  use BottomSheetBackdrop component and on some devices when the screen is mounted, no button works on that screen, it feels like something is overlayed on top of that. This bug only exists on android's some devices.  
For example, it doesn't work: on Huawei mate 8 version 7.0,  Nokia 2.1 version 10, Samsung galaxy s10 version 10, techno camon 15 version 10. 

On the other hand, I tested it on my phones: Meizu note 5 version 7.0, Google pixel 3 version 10 and everything works perfectly( as I said even IOS doesn't have such a problem ). 

Thanks.
## Environment info

| Library                         | Version |
| ------------------------------- | ------- |
| @gorhom/bottom-sheet         | ^3.6.6   |
| react-native                    | 0.64.1   |
| react-native-reanimated         | ^2.2.0   |
| react-native-gesture-handler    | ^1.10.3   |

## Steps To Reproduce

1. Run codes
2. Try to touch anything on the screen when you use this library

Describe what you expected to happen:

1. screen should be clickable.

## Reproducible sample code
```jsx
function ScreenComponent() {
  const bottomSheetRef = useRef<BottomSheet>(null);
  const [isInsightActive, setIsInsightActive] = useState(true);

  const { bottom } = useSafeAreaInsets();

  const onIsItHelpfulPress = useCallback(() => {
    bottomSheetRef.current?.expand();
  }, []);

  return (
    <Screen stretch>
      <Header withTitle />
      <Row justifySpaceBetween alignCenter px={7} mt={7}>
        <Touchable active={isInsightActive} onPress={onIsItHelpfulPress}>
          <Text fontSize={4} fontWeight={3}>
            Insights
          </Text>
        </Touchable>
      </Row>

      <BottomSheet
        ref={bottomSheetRef}
        index={-1}
        snapPoints={[0, 200]}
        backdropComponent={BottomSheetBackdrop}
      >
        <Column mt={7} stretch paddingBottom={`${bottom}px`}>
          <Text fontSize={5} fontWeight={3} px={7}>
            Is this helpful?
          </Text>
        </Column>
      </BottomSheet>
    </Screen>
  );
}
```

**Comment by harsh-siriah** [comment]
I was facing a similar issue. 
My workaround was to put 0 as the first element in my snap points array and then pass index={0} in the <BottomSheet> component.

`const snapPoints = [0, '50%']` 

` <BottomSheet index={0}>`

**Comment by beqramo** [comment]
Thanks I will try when I get on my laptop, but their doc definitely says that you should pass -1 if you want to start with closed mode. 

https://gorhom.github.io/react-native-bottom-sheet/props#index

It seems like a bug for me.

**Comment by beqramo** [comment]
@harshsiri110  that workaround worked thanks, but, still, I will leave it open. it needs fixes

**Comment by harsh-siriah** [comment]
> @harshsiri110 that workaround worked thanks, but, still, I will leave it open. it needs fixes

It's great that it worked! I am facing the same issue in some devices even after the workaround though. I hope the team fixes this soon.

**Comment by whalemare** [comment]
The same issue, but when we pass snap points started from 0, like `[0, 200]`, BottomSheet can be closed.

In my case, `BottomSheet` is always opened in state like `['50%'; '100%']`

**Comment by gorhom** [comment]
hi guys, could you test this issue on `v4` and let me know if its occurs, thanks !

**Comment by whalemare** [comment]
I am use `4.0.0-alpha.27` and issue still exists, but not all the time. Only in some specific cases, I dont know why

**Comment by soutot** [comment]
Try using TouchableOpacity from the library (`import { TouchableOpacity } from '@gorhom/bottom-sheet'`) instead of using the native. It worked for me

**Comment by whalemare** [comment]
> Try using TouchableOpacity

It is not about this issue. 
Whole screen where placed BottomSheet not clickable, includes content outside of BottomSheet. Only restart app solves issue sometimes

One more, if we call `Debug menu` it will appear and clickable. I was trying to enable monitoring views and root tree, but it enables only after click on content, but whole content not clickable

**Comment by Ripul01** [comment]
I have encountered the same issue on android while using the physical back button.

So to recreate the bug every time all I have to do is this:

https://user-images.githubusercontent.com/48410268/130329473-6e8e9d2e-db1e-4c89-afa4-9a606c1ac67a.gif

Basically, If I open the keyboard while on the bottom sheet and close the keyboard with the physical back button and subsequently close the bottom sheet with the physical back button it renders the whole app unresponsive.

Not sure what the issue is but I did try doing the same thing everywhere in the app resulting in the same bug.

If any has any idea help would be greatly appreciated

**Comment by whalemare** [comment]
@Ripul01 can you provide a reproducible code sample, in expo for example?

**Comment by Ripul01** [comment]
Yeah sure, I'll make it tomorrow and add it here

**Comment by Ripul01** [comment]
@whalemare Here is the link to the test project: https://github.com/Ripul01/BottomSheetUnresponsiveBug.

Here I am facing a new issue now. The app does not seem to go unresponsive but if you open the keyboard, then close the keyboard then the modal, then the modal does not want to open.

So adding the other modal works fine by itself but if you reproduce the bug in the first modal the second modal will not close on the physical back button press.

Here is another gif to see that in action:

1) All modals work fine by themselves: (https://user-images.githubusercontent.com/48410268/130331890-32a2474e-4234-45de-ae6d-b9c481954d8b.gif)

2) Single Modal does not open upon reproducing the bug: (https://user-images.githubusercontent.com/48410268/130331934-3c69bb9f-aa59-4b6d-9c97-73fd97ffc0ff.gif)

3) Second Modal does not close after producing the bug on the first modal: (https://user-images.githubusercontent.com/48410268/130331942-2051141f-d6c2-4e38-a9c7-eb85420c5ad6.gif)

**Comment by gorhom** [comment]
@Ripul01 could you use `v4` ?

**Comment by Ripul01** [comment]
@gorhom Yeah sure. I’ll make a bare react native project because it defaults to 3.6.6 in expo. I will post it tomorrow

**Comment by gorhom** [comment]
@Ripul01 you can use this snack https://snack.expo.dev/@gorhom/bottom-sheet-v4-reproducible-issue-template

**Comment by Ripul01** [comment]
@gorhom Hey so the snack template has the bottom sheet and not the bottom sheet modal used and if I switch it up to the bottom sheet modal it just doesn't work for some reason. Would it be alright if I made a bare react-native project instead of an expo one on a snack?

**Comment by Ripul01** [comment]
@gorhom I tried reproducing the issue with the 4.0.0-alpha.29 and I was not able to do it. It is working properly and is not giving out weird bugs. I also tried it on my other project which also works fine but it broke my dynamic snap point setup. Has the setup for dynamic snap points changed in the new version. Earlier I was managing the modal height as state initialized as 0 which was overwritten by on layout from the bottomsheetview by the actual modal height but it now gives the error that snap point 0 is invalid.

The example code for the dynamic snap point also doesn't use the new version I suppose because that also does not seem to work.

**Comment by gorhom** [comment]
@Ripul01 i have updated the dynamic snap point handling , look at the example below

https://github.com/gorhom/react-native-bottom-sheet/blob/master/example/src/screens/advanced/DynamicSnapPointExample.tsx

**Comment by Ripul01** [comment]
@gorhom Thank you. It seems everything is working now. The issue of weird freezing was gone after using the latest build and the dynamic snap point example helped a lot.

**Comment by beqramo** [comment]
I will close it then

**Comment by Zakyyy** [comment]
It is happening with me, after I dismiss the modal and tries to open it again (using present) nothing happens

**Comment by Evan-Younan** [comment]
It looks like I encountered this issue on `4.0.3`. This will happen on some Android devices but not others. The Samsung S10 reproduces this bug 100% of the time for example. Code looks like so:

```
 <BottomSheet
    ref={bottomModalSheetRef}
    index={-1}
    animationConfigs={animationConfigs}
    snapPoints={["55%"]}
    detached={false}
    enablePanDownToClose={true}
    enableContentPanningGesture={true}
    overDragResistanceFactor={4}
    topInset={insets.top}
    backdropComponent={(bdProps) => (
      <BottomSheetBackdrop
          {...bdProps}
          appearsOnIndex={0}
          disappearsOnIndex={-1}
          opacity={0.4}
          enableTouchThrough
      />
    )}
    backgroundStyle={{ backgroundColor: "white" }}
>
  <BottomSheetView style={styles.container}>
    <SortPickerContent
       setSortType={setSortType}
       sortType={sortType}
       handleBottomSheetClose={handleBottomSheetClose}
    />
  </BottomSheetView>
</BottomSheet>


...


const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.white,
  },
});

```

edit: After some more digging, a lot of the issues are happening with Android 10 devices.

**Comment by fifi98** [comment]
I'm having the same issue with the latest 4.1.0 version when using dynamic snap points. It works perfectly fine without dynamic snap points.

**Comment by ardyfeb** [comment]
I'm having same issue, flatlist not rendering

**Comment by eightyfive** [comment]
On Android, the Reanimated tap gesture apparently does not work by default. You have to wrap your app with `<GestureHandlerRootView style={{ flex: 1 }}>`.

This is why the `BottomSheetBackdrop` component does not call "close" on press.

Strangely enough, this is not needed for iOS.

_Source_: https://github.com/software-mansion/react-native-reanimated/discussions/2794

**Comment by Cherub12** [comment]
I found a fix:
Go to node_modules > @gorhom > bottom-sheet > src > components > BottomSheet > BottomSheetBackdrop and replace this line under 
//#region effects
  useAnimatedReaction(
    () => !(animatedIndex.value >= appearsOnIndex),
    (shouldDisableTouchability, previous) => {
      if (shouldDisableTouchability === previous) {
        return;
      }
      runOnJS(handleContainerTouchability)(shouldDisableTouchability);
    },
    [appearsOnIndex]
  );

**Comment by sreejithsivaji** [comment]
disappearsOnIndex={-1},  I just commented out this prop in the BottomSheetBackdrop. 

// Default backdrop component
    const renderBackdrop = useCallback(
      (props: BottomSheetBackdropProps) => (
        <BottomSheetBackdrop
          {...props}
          // disappearsOnIndex={-1}, ---------- not working when it's enabled! ----------
          appearsOnIndex={0}
          opacity={0.5}
          enableTouchThrough={false}
        />
      ),
      []
    );


// ---------------- full code  ----------------

// BottomSheet component interface
export interface BottomSheetRef {
  expand: () => void;
  close: () => void;
  collapse: () => void;
  snapToIndex: (index: number) => void;
  snapToPosition: (position: string | number) => void;
}

interface BottomSheetProps {
  children: React.ReactNode;
  snapPoints?: (string | number)[];
  enablePanDownToClose?: boolean;
  backdropComponent?: React.FC<BottomSheetBackdropProps>;
  title?: string;
  headerBorder?: boolean,
  onSheetChange?: (index: number) => void;
  backgroundStyle?: object;
  handleStyle?: object;
  enableDynamicSizing?: boolean;
  initialIndex?: number;
  defaultOpen?: boolean;
  defaultSnapIndex?: number;
}

// Main BottomSheet Component
const BottomSheet = forwardRef<BottomSheetRef, BottomSheetProps>(
  (
    {
      children,
      snapPoints = ["25%", "50%", "90%"],
      enablePanDownToClose = true,
      backdropComponent,
      title,
      headerBorder = true,
      onSheetChange,
      backgroundStyle,
      handleStyle,
      enableDynamicSizing = false,
      initialIndex = -1,
      defaultOpen = false,
      defaultSnapIndex,
    },
    ref
  ) => {
    const bottomSheetRef = useRef<BottomSheetLib>(null);
    const { colorScheme } = useColorScheme();

    // Memoize snap points
    const snapPointsMemo = useMemo(() => snapPoints, [snapPoints]);

    // Calculate the initial index based on default props
    const calculatedInitialIndex = useMemo(() => {
      // If initialIndex is explicitly provided (including -1 or 0), use it
      if (initialIndex !== undefined && initialIndex !== null) {
        return initialIndex;
      }
      if (defaultOpen) {
        // If defaultSnapIndex is explicitly provided (including 0), use it
        // Otherwise use the highest snap point
        return defaultSnapIndex !== undefined && defaultSnapIndex !== null
          ? defaultSnapIndex
          : snapPoints.length - 1;
      }

      // Sheet is closed by default
      return -1;
    }, [defaultOpen, defaultSnapIndex, snapPoints.length]);

    // Expose methods through ref
    useImperativeHandle(ref, () => ({
      expand: () => {
        if (defaultSnapIndex !== undefined && defaultSnapIndex !== null) {
          bottomSheetRef.current?.snapToIndex(defaultSnapIndex);
        } else {
          bottomSheetRef.current?.expand();
        }
      },
      close: () => bottomSheetRef.current?.close(),
      collapse: () => bottomSheetRef.current?.collapse(),
      snapToIndex: (index: number) =>
        bottomSheetRef.current?.snapToIndex(index),
      snapToPosition: (position: string | number) =>
        bottomSheetRef.current?.snapToPosition(position),
    }));


    // Handle sheet changes
    const handleSheetChanges = useCallback(
      (index: number) => {
        onSheetChange?.(index);
      },
      [onSheetChange]
    );

    // Default backdrop component
    const renderBackdrop = useCallback(
      (props: BottomSheetBackdropProps) => (
        <BottomSheetBackdrop
          {...props}
          // disappearsOnIndex={-1}, ---------- not working when it's enabled! ----------
          appearsOnIndex={0}
          opacity={0.5}
          enableTouchThrough={false}
        />
      ),
      []
    );

    return (
      <BottomSheetLib
        ref={bottomSheetRef}
        index={calculatedInitialIndex}
        snapPoints={snapPointsMemo}
        onChange={handleSheetChanges}
        enablePanDownToClose={enablePanDownToClose}
        backdropComponent={backdropComponent || renderBackdrop}
        handleIndicatorStyle={{
          backgroundColor: colorScheme === "dark" ? "#fff" : "#e5e5e5",
          width: 40,
          height: 4,
          ...handleStyle,
        }}
        backgroundStyle={{
          borderTopLeftRadius: 24,
          borderTopRightRadius: 24,
          elevation: 10,
          backgroundColor:
            colorScheme === "dark"
              ? ThemeColors.background_dark
              : ThemeColors.background,
          ...backgroundStyle,
        }}
        enableDynamicSizing={enableDynamicSizing}
      >
        {title && (
          <View className="px-6">
            <View className={`pb-4 ${headerBorder ? "border-b border-border dark:border-border-dark drop-shadow" : "" }`}>
              <Text className="font-sans-semibold text-xl text-foreground dark:text-foreground-dark text-center">
                {title}
              </Text>
            </View>
          </View>
        )}
        <BottomSheetScrollView
          contentContainerStyle={{
            paddingHorizontal: 24,
            paddingTop: 10,
            paddingBottom: 24,
          }}
          showsVerticalScrollIndicator={false}
          className="bg-background dark:bg-background-dark"
        >
          {children}
        </BottomSheetScrollView>
      </BottomSheetLib>
    );
  }
);

BottomSheet.displayName = "BottomSheet";

export default BottomSheet;
